### ISSUE TITLE
Implement indexed access on Window (window.frames).

### ISSUE THREAD

**Issue by Ms2ger** [servo/servo#4589]
I believe this is implemented on WindowProxy in Gecko; @bzbarsky knows more.

**Comment by Ms2ger** [comment]
Blocks #2968, #2185.

**Comment by bzbarsky** [comment]
That's correct.  You can see the implementation in the various proxy handler methods on nsOuterWindowProxy in Gecko.

This is, in fact, the only sane way to implement it without major spidermonkey changes, since you want the thing that has these indexed props to be a proxy and SpiderMonkey doesn't support proxies as globals.

**Comment by evilpie** [comment]
I will try to implement parts of the WindowProxy based on the nsOuterWindowProxy in Gecko.

**Comment by Ms2ger** [comment]
What remains to be done is implementing `IndexedGetter` on Window; it always returns `None` now.

**Comment by benschulz** [comment]
I've [implemented IndexedGetter](https://github.com/benschulz/servo/commit/a01b0776ac210c927a3dd1c5900736e810b444ad) and it works as expected. However, the finer points of the specification aren't met. Specifically servo is missing gecko's outer+inner window concept. What is there is essentially the inner window. This has two implications w.r.t. to the [specified](https://html.spec.whatwg.org/multipage/browsers.html#dom-window-item) order.

 - The initial order depends on the order in which frames' documents are loaded. Also, the window is not accessible until it was asynchronously loaded (#5500). 
 - Whenever a frame's document changes, its browsing context is moved to the end of the list. (Its `Page` is removed from `page.children` and a new one pushed onto it).

**TL;DR**: The initial order may be wrong and it may incorrectly change on top of that. To fix this, something like geckos outer+inner window pairing is needed. I would like to work on that, unless there are plans for some other solution.

**Comment by bzbarsky** [comment]
You need the inner/outer thing for basic web compat anyway, so I doubt there should be any other plans...

**Comment by jdm** [comment]
We do actually have an inner/outer concept already, it's just incorrect. That's tracked in #5238.

**Comment by benschulz** [comment]
Naming-wise that makes a lot of sense; `BrowserContext` should implement the [browsing context concept](https://html.spec.whatwg.org/multipage/browsers.html#browsing-context) (which happens to be the equivalent of gecko's outer windows), correct?

Either way, I'm gonna look into this. I'm experienced with neither C++ nor Rust, so I hesitate to have one of these issues assigned to me. Maybe in a couple of days or so… ;)

**Comment by jdm** [comment]
Correct!

**Comment by ecoal95** [comment]
Another test affected by this is `cssom-view/elementsFromPoint.html`.

So... This blocks #10091

**Comment by KiChjang** [comment]
We already have `get_property_descriptor` defined in `script/dom/bindings/proxyhandle.rs` and hooked up with WindowProxy traps, but I believe the reason why it doesn't work as expected nowadays is due to #11868.
